appsrc/dcfile/dctoraw.man
.TH DCTORAW 1 "6 March 2014" "DICOM PS3" "DICOM PS3 - DICOM image to raw file"
.SH NAME
dctoraw \- ACR/NEMA DICOM PS3 ... DICOM PS3 - DICOM image to raw file
.SH SYNOPSIS
.HP 10
.B dctoraw
.so man1/gen.so
[
.B \-v|verbose
]
[
.B \-quiet|silent
]
.so man1/optin.so
.so man1/genout.so
.SH DESCRIPTION
.LP
.B dctoraw
reads the named dicom or acr-nema input file and copies the raw image
pixel data to a raw binary file without a header of any kind.
.LP
The byte order, packing or encapsulation of the raw result is dependent only
on the encoding of the input file and cannot be changed. Use dccp(1)
first to change the encoding if necessary.
.LP
If the transfer syntax is not encapsulated, only the necessary number of bytes
to encode the pixel data will be written (based on rows, columns, number of
frames, samples per pixel and bits allocated); this means that for an odd number
of bytes the extra DICOM padding byte will not be copied to the raw output. If
the transfer syntax is encapsulated, or any of the necessary attributes to calculate
the correct pixel data length are missing, all encoded bytes will be written.
.LP
.SH OPTIONS
The raw output goes where it is specified or standard out.
The verbose output goes to standard error.
.LP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-quiet|silent
.RS
Suppress the normal description of image parameters.
.RE
.TP
.B \-verbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents while reading, once read, during replacement, and before writing.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% dctoraw expmeta.dc3 expmeta.raw
.RE
******** Parameters ... ********
.RE
(0x0028,0x0010) US Rows                       [0x0100] 
.RE
(0x0028,0x0011) US Columns                    [0x0100] 
.RE
(0x0028,0x0002) US Samples per Pixel          [0x0001] 
.RE
(0x0028,0x0103) US Pixel Representation       [0x0000] 
.RE
(0x0028,0x0004) CS Photometric Interpretation <MONOCHROME2> 
.RE
(0x0028,0x0100) US Bits Allocated             [0x0010] 
.RE
(0x0028,0x0101) US Bits Stored                [0x0010] 
.RE
(0x0028,0x0102) US High Bit                   [0x000f] 
.RE
Encapsulated    No
.RE
ByteOrder       Little
.RE
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dccp(1) ,
.BR pnmtodc(1) ,
.BR rawtodc(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
241220:	libsrc/src/dctool/attrtypv.cc: Allow TAB character in control characters for text VRs (CP 1425)

241114:	appsrc/dcfile/dciodvfy.cc: Allow single byte of padding for odd number of 8 bit LUT entries in 16 bit words (000607)

241114:	appsrc/dcfile/dciodvfy.cc: Remove check on SegmentNumber monotonically increasing from 1 (not 0) if SegmentationType is LABELMAP (Sup 243)

241111:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Ion Range Surface Compensator (CP 2401), Isodose levels for RT Dose visualization (CP 2402)

241105:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Ocular Region Imaged Module in Ophthalmic Tomography (CP 2346)

241003:	libsrc/standard/condn.tpl,module/base.tpl: Restore not requiring SegmentationMacro if TILED_FULL (CP 2331) lost when adding Label Map Segmentation

240915:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Add SOP Class and data elements for Height Map Segmentation and Revised Ophthalmic OCT En Face Image (Sup 240)

240914:	libsrc/standard/condn.tpl,sopcl.tpl,iodcomp,module,strval/base.tpl: Add Label Map Segmentation, and remove related experimental profile (Sup 243)

240913:	libsrc/standard/elmdict/dicom3.tpl: Add data element for Clarify Description related to Range Shifter Settings reusing DICONDE Material Density (CP 2373 FT2)

240912:	libsrc/standard/elmdict/dicom3.tpl: Add data element for Differentiate acquisition techniques for RT 3D CBCT imaging (CP 2220)

240822: libsrc/standard/module/base.tpl:	Correctly Add Ethnic Group Code Sequence (CP 2356)

240728: libsrc/standard/condn.tpl,module/base.tpl: Report error when PositionReferenceIndicator not SLIDE_CORNER for microscopy (000605)

240712:	appsrc/dcfile/dcacqmap.script,dcacqmap.man: Fix invalid character range in egrep for position files (000604), typo in documentation

240623:	libsrc/standard/elmdict/diconde.tpl: Add DICONDE data elements for thermography imaging

240612:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Table Top Position Alignment UID to Table Top positions (CP 2376)

240612:	libsrc/standard/elmdict/dicom3.tpl: Add data elements for Improve RT equipment mapping for imaging (CP 2374)

240611:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl:	Add Ethnic Group Code Sequence (CP 2356)

240610:	libsrc/standard/elmdict/dicom3.tpl:	Add data element for updated RT Patient Setup Module (CP 2344)

240515:	libsrc/standard/condn.tpl,module/base.tpl: Check ICCProfile and ColorSpace not present in top level dataset Image Pixel Module if Optical Path Module present (000603)

240506:	libsrc/standard/module/vl.tpl: Check AnnotationGroupNumber is not zero (000602)

240324:	libsrc/standard/condn.tpl,iodcomp/base.tpl: Detect FrameOfReference Module if FoR UID present even if position and orientation not, since MBPO in Secondary Capture IOD (CP 2330)

240402:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check AnnotationGroupUID is not duplicated (000601)

240329:	libsrc/standard/elmdict/other.tpl: 3DHISTECH private data elements from conformance statement

240329:	appsrc/dcfile/dcentvfy.cc,libsrc/src/locale/mesgtext.tpl: Check all ReferencedSOPInstanceUID instances are in set of supplied files (000600)

240328:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check sanity of Microscopy Bulk Simple Annotation indices and coordinates (000599)

240328:	libsrc/include/dctool/attr.h,attrtype.h,attrtypo.h,src/dctool/attrmxrd.cc,attrtypo.cc: Add support of in-memory (small) OL VR data elements to get values to check (000599)

240325:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Clinical Trial Identifiers need Issuers and may be multiple (CP2335)

240324:	libsrc/standard/condn.tpl,iodcomp,module/base.tpl: Clarify Segment scan order for TILED_FULL (CP 2331) and remove CP2331 profile option to test TILED_FULL segmentations

240324:	libsrc/standard/condn.tpl,iodcomp/base.tpl: Add Spatial Information to Secondary Capture IOD (CP 2330)

240324:	libsrc/standard/elmdict/dicom3.tpl,module/vl.tpl: Add information about Vertex Distance to refraction correction parameters (CP 2324)

240324:	libsrc/standard/elmdict/dicom3.tpl,module/base.tpl: Add Synthetic Data Flag (CP 2320)

240316:	appsrc/dcfile/dcentvfy.cc: Check Frame of Reference IE attribute values same for same Series (000597)

240312:	libsrc/standard/elmdict/philips.tpl: Add more observed private data elements

240224:	appsrc/dcfile/dciodvfy.cc,libsrc/include/dctool/attr.h,attrothr.h,transyn.h,src/dctool/transyn.cc,transyne.h,locale/mesgtext.tpl,standard/transyn.tpl,support/transyn.awk: Add check that Basic Offset Table is valid if present, including length, point to Item Tag and begin with a JPEG/J2K SOI marker (000531) and point to even offsets (000526)

240222:	appsrc/acrnema/andump.cc,libsrc/include/dctool/attr.h,attrtype.h,attrtypn.h,attrtypo.h,dcstream.h,vr.h,generic/bnstream.h,src/dctool/attr.cc,attrmxrd.cc,attrnew.cc,attrtypo.cc: Add 64 bit binary VRs (CP 1819)

240215:	libsrc/standard/condn.tpl,module/vl.tpl: Correct spurious report of invalid CommonZCoordinateValue for nnotationCoordinateType not 3D when not actually present (000596)

240214:	appsrc/dcfile/dcentvfy.cc: Check Equipment IE attribute values same for same Series (000595)

240131:	libsrc/standard/condn.tpl,iodcomp,module,strval/base.tpl: Add profile option to test LABELMAP segmentations (Sup243)

240118:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: Check that SCOORD3D POLYGON are closed (first and last points same) (000592)

240114:	libsrc/standard/elmdict/dicom3.tpl: Provide additional ROI parameters (CP 2296)

240112:	libsrc/standard/elmdict/dicom3.tpl: Add new data elements for DICOMweb Storage Commitment Service (Sup 234)

231125:	config/Configure: Build two architectures when clang >=15 but not with ld flag "-force_cpusubtype_ALL" that is no longer supported

231115:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Add Confocal Microscopy data elements and SOP Class UIDs (Sup 236)


appsrc/dcfile/dctoraw.cc
static const char *CopyrightIdentifier(void) { return "@(#)dctoraw.cc Copyright (c) 1993-2024, David A. Clunie DBA PixelMed Publishing. All rights reserved."; }
#include "attrmxls.h"
#include "attrothr.h"
#include "attrval.h"
#include "mesgtext.h"
#include "ioopt.h"
#include "dcopt.h"
#include "elmconst.h"
#include "transynd.h"

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);
	OutputOptions		output_options(options);

	bool verbose=options.get("verbose") || options.get("v");
	bool quiet=options.get("quiet") || options.get("silent");

	dicom_input_options.done();
	output_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);
	OutputOpenerFromOptions output_opener(
		options,output_options.filename,cout);

	cerr << dicom_input_options.errors();
	cerr << output_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();
	cerr << output_opener.errors();

	if (!dicom_input_options.good()
	 || !output_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !output_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< output_options.usage()
			<< " [-v|-verbose]"
			<< " [-quiet|-silent]"
			<< " [" << MMsgDC(InputFile)
				<< "[" << MMsgDC(OutputFile) << "]]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);
	ostream out(output_opener);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream log(cerr);
	if (verbose) log << "******** While reading ... ********" << endl; 
	list.read(din,false/*newformat*/,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	const char *errors=list.errors();
	if (errors) log << errors << flush;
	if (!list.good()) {
		log << EMsgDC(DatasetReadFailed) << endl;
		success=false;
	}
	
	// Most of this is copied from dctopnm ... should refactor :(

	Uint16 vRows = 0;
	Attribute *aRows = list[TagFromName(Rows)];
	if (!aRows)
		log << WMsgDC(MissingAttribute)
		    << " - \"Rows\""
		    << endl;
	else
		vRows=AttributeValue(aRows);

	Uint16 vColumns = 0;
	Attribute *aColumns = list[TagFromName(Columns)];
	if (!aColumns)
		log << WMsgDC(MissingAttribute)
		    << " - \"Columns\""
		    << endl;
	else
		vColumns=AttributeValue(aColumns);

	Uint16 vNumberOfFrames = 0;
	Attribute *aNumberOfFrames = list[TagFromName(NumberOfFrames)];
	if (aNumberOfFrames)	// optional
		vNumberOfFrames=AttributeValue(aNumberOfFrames);

	if (vNumberOfFrames == 0) {
		vNumberOfFrames = 1;	// need to treat it as 1 if missing for later byteCount calculation
	}

	char *vPhotometricInterpretation = 0;
	Attribute *aPhotometricInterpretation = list[TagFromName(PhotometricInterpretation)];
	if (!aPhotometricInterpretation)
		log << WMsgDC(MissingAttribute)
		    << " - \"PhotometricInterpretation\""
		    << endl;
	else
		vPhotometricInterpretation=AttributeValue(aPhotometricInterpretation);

	Uint16 vSamplesPerPixel = 0;
	Attribute *aSamplesPerPixel = list[TagFromName(SamplesPerPixel)];
	if (!aSamplesPerPixel)
		log << WMsgDC(MissingAttribute)
		    << " - \"SamplesPerPixel\""
		    << endl;
	else
		vSamplesPerPixel=AttributeValue(aSamplesPerPixel);

	Uint16 vBitsAllocated = 0;
	Attribute *aBitsAllocated = list[TagFromName(BitsAllocated)];
	if (!aBitsAllocated)
		log << WMsgDC(MissingAttribute)
		    << " - \"BitsAllocated\""
		    << endl;
	else
		vBitsAllocated=AttributeValue(aBitsAllocated);

	Uint16 vBitsStored = 0;
	Attribute *aBitsStored = list[TagFromName(BitsStored)];
	if (!aBitsStored)
		log << WMsgDC(MissingAttribute)
		    << " - \"BitsStored\""
		    << endl;
	else
		vBitsStored=AttributeValue(aBitsStored);

	Uint16 vHighBit = 0;
	Attribute *aHighBit = list[TagFromName(HighBit)];
	if (!aHighBit)
		log << WMsgDC(MissingAttribute)
		    << " - \"HighBit\""
		    << endl;
	else
		vHighBit=AttributeValue(aHighBit);

	Uint16 vPixelRepresentation = 0xffff;
	Attribute *aPixelRepresentation = list[TagFromName(PixelRepresentation)];
	if (!aPixelRepresentation)
		log << WMsgDC(MissingAttribute)
		    << " - \"PixelRepresentation\""
		    << endl;
	else
		vPixelRepresentation=AttributeValue(aPixelRepresentation);

	Uint16 vPlanarConfiguration = 0xffff;
	Attribute *aPlanarConfiguration = list[TagFromName(PlanarConfiguration)];
	if (vSamplesPerPixel > 1 && !aPlanarConfiguration)
		log << WMsgDC(MissingAttribute)
		    << " - \"PlanarConfiguration\""
		    << endl;
	else
		vPlanarConfiguration=AttributeValue(aPlanarConfiguration);

	Uint32 byteCount = ((Uint32)vRows) * vColumns * vNumberOfFrames * vSamplesPerPixel * ((vBitsAllocated-1)/8 + 1);

	TransferSyntax *ts=din.getTransferSyntaxToReadDataSet();
	Assert(ts);

	Attribute *apixeldata=list[TagFromName(PixelData)];
	if (!apixeldata) {
		log << EMsgDC(MissingAttribute) << " - \"PixelData\"" << endl;
		success=false;
	}
	else if (!apixeldata->isOtherData()) {
		log << EMsgDC(PixelDataIncorrectVR) << endl;
		success=false;
	}
	else {
		if (!quiet) {
			// this is now redundant with the explicit retrieval of values above, but keep this redundant dump in case anyone used these in scripts ...
			log << "******** Parameters ... ********" << endl; 
			ElementDictionary *dict=list.getDictionary();
			Attribute *a;
			a=list[TagFromName(NumberOfFrames)];
			if (a) { a->write(log,dict); log << endl; }
			a=list[TagFromName(Rows)];
			if (a) { a->write(log,dict); log << endl; }
			a=list[TagFromName(Columns)];
			if (a) { a->write(log,dict); log << endl; }
			a=list[TagFromName(SamplesPerPixel)];
			if (a) { a->write(log,dict); log << endl; }
			a=list[TagFromName(PixelRepresentation)];
			if (a) { a->write(log,dict); log << endl; }
			a=list[TagFromName(PhotometricInterpretation)];
			if (a) { a->write(log,dict); log << endl; }
			a=list[TagFromName(PlanarConfiguration)];
			if (a) { a->write(log,dict); log << endl; }
			a=list[TagFromName(BitsAllocated)];
			if (a) { a->write(log,dict); log << endl; }
			a=list[TagFromName(BitsStored)];
			if (a) { a->write(log,dict); log << endl; }
			a=list[TagFromName(HighBit)];
			if (a) { a->write(log,dict); log << endl; }

			log << "\tRows = " << dec << vRows << endl;
			log << "\tColumns = " << dec << vColumns << endl;
			log << "\tNumberOfFrames = " << dec << vNumberOfFrames << endl;
			log << "\tPhotometricInterpretation = "
				<< (vPhotometricInterpretation ? vPhotometricInterpretation : "")
				<< endl;
			log << "\tSamplesPerPixel = " << dec << vSamplesPerPixel << endl;
			log << "\tBitsAllocated = " << dec << vBitsAllocated << endl;
			log << "\tBitsStored = " << dec << vBitsStored << endl;
			log << "\tHighBit = " << dec << vHighBit << endl;
			log << "\tPixelRepresentation = " << dec << vPixelRepresentation << endl;
			log << "\tPlanarConfiguration = " << hex << vPlanarConfiguration << dec << endl;
			log << "\tRows*Columns*vNumberOfFrames*SamplesPerPixel*((vBitsAllocated-1)/8+1) = "
				<< hex << byteCount
				<< dec << " (" << byteCount << " dec)" << endl;

			dumpTransferSyntaxEncapsulation(ts);
			dumpTransferSyntaxByteOrder(ts);
		}

		OtherUnspecifiedLargeAttributeBase *
			opixeldata = apixeldata->castToOtherData();
		Assert(opixeldata);

		if (ts->isEncapsulated()) {
			if (!quiet) log << "Writing encapsulated Transfer Syntax ..." << endl;
			opixeldata->writeRaw(out);
		}
		else if (byteCount == 0) {
			// do it the "old" way, if for some reason we could not find a suitable value for something that contributes to the desired count; should never happen
			if (!quiet) log << "Writing VL " << dec << opixeldata->getVL() << " (dec) bytes ..." << endl;
			opixeldata->writeRaw(out);
		}
		else {
			if (!quiet) log << "Writing unpadded " << dec << byteCount << " (dec) bytes ..." << endl;
			opixeldata->writeRaw(out,0/*byte offset*/,byteCount);
		}
		
		if (!out.good()) {
			log << EMsgDC(Writefailed) << endl;
			success=false;
		}
	}

	return success ? 0 : 1;
}

	
